File: agentauth-sdk-python/agentauth_sdk/permissions.py
"""
Type-safe permission system for AgentAuth SDK

Provides constants and type hints for all service:resource:action permissions.
"""

from typing import Literal, Union

# Type definitions for permissions
Permission = Union[
    # Zendesk
    Literal["zendesk:tickets:read"],
    Literal["zendesk:tickets:write"],
    Literal["zendesk:users:read"],
    Literal["zendesk:users:write"],
    Literal["zendesk:tickets:*"],
    Literal["zendesk:users:*"],
    Literal["zendesk:*:*"],
    # Slack
    Literal["slack:messages:read"],
    Literal["slack:messages:write"],
    Literal["slack:channels:read"],
    Literal["slack:channels:write"],
    Literal["slack:messages:*"],
    Literal["slack:channels:*"],
    Literal["slack:*:*"],
    # HubSpot
    Literal["hubspot:contacts:read"],
    Literal["hubspot:contacts:write"],
    Literal["hubspot:deals:read"],
    Literal["hubspot:deals:write"],
    Literal["hubspot:companies:read"],
    Literal["hubspot:contacts:*"],
    Literal["hubspot:deals:*"],
    Literal["hubspot:companies:*"],
    Literal["hubspot:*:*"],
    # GitHub
    Literal["github:repos:read"],
    Literal["github:repos:write"],
    Literal["github:issues:read"],
    Literal["github:issues:write"],
    Literal["github:pull_requests:read"],
    Literal["github:repos:*"],
    Literal["github:issues:*"],
    Literal["github:pull_requests:*"],
    Literal["github:*:*"],
    # Salesforce
    Literal["salesforce:accounts:read"],
    Literal["salesforce:accounts:write"],
    Literal["salesforce:leads:read"],
    Literal["salesforce:leads:write"],
    Literal["salesforce:accounts:*"],
    Literal["salesforce:leads:*"],
    Literal["salesforce:*:*"],
    # Stripe
    Literal["stripe:payments:read"],
    Literal["stripe:customers:read"],
    Literal["stripe:invoices:read"],
    Literal["stripe:payments:*"],
    Literal["stripe:customers:*"],
    Literal["stripe:invoices:*"],
    Literal["stripe:*:*"],
    # Admin
    Literal["*:*:*"],
]


class _ZendeskTickets:
    """Zendesk Tickets permissions"""

    Read: Permission = "zendesk:tickets:read"
    Write: Permission = "zendesk:tickets:write"
    All: Permission = "zendesk:tickets:*"


class _ZendeskUsers:
    """Zendesk Users permissions"""

    Read: Permission = "zendesk:users:read"
    Write: Permission = "zendesk:users:write"
    All: Permission = "zendesk:users:*"


class _Zendesk:
    """Zendesk service permissions"""

    Tickets = _ZendeskTickets()
    Users = _ZendeskUsers()
    All: Permission = "zendesk:*:*"


class _SlackMessages:
    """Slack Messages permissions"""

    Read: Permission = "slack:messages:read"
    Write: Permission = "slack:messages:write"
    All: Permission = "slack:messages:*"


class _SlackChannels:
    """Slack Channels permissions"""

    Read: Permission = "slack:channels:read"
    Write: Permission = "slack:channels:write"
    All: Permission = "slack:channels:*"


class _Slack:
    """Slack service permissions"""

    Messages = _SlackMessages()
    Channels = _SlackChannels()
    All: Permission = "slack:*:*"


class _HubSpotContacts:
    """HubSpot Contacts permissions"""

    Read: Permission = "hubspot:contacts:read"
    Write: Permission = "hubspot:contacts:write"
    All: Permission = "hubspot:contacts:*"


class _HubSpotDeals:
    """HubSpot Deals permissions"""

    Read: Permission = "hubspot:deals:read"
    Write: Permission = "hubspot:deals:write"
    All: Permission = "hubspot:deals:*"


class _HubSpotCompanies:
    """HubSpot Companies permissions"""

    Read: Permission = "hubspot:companies:read"
    All: Permission = "hubspot:companies:*"


class _HubSpot:
    """HubSpot service permissions"""

    Contacts = _HubSpotContacts()
    Deals = _HubSpotDeals()
    Companies = _HubSpotCompanies()
    All: Permission = "hubspot:*:*"


class _GitHubRepos:
    """GitHub Repositories permissions"""

    Read: Permission = "github:repos:read"
    Write: Permission = "github:repos:write"
    All: Permission = "github:repos:*"


class _GitHubIssues:
    """GitHub Issues permissions"""

    Read: Permission = "github:issues:read"
    Write: Permission = "github:issues:write"
    All: Permission = "github:issues:*"


class _GitHubPullRequests:
    """GitHub Pull Requests permissions"""

    Read: Permission = "github:pull_requests:read"
    All: Permission = "github:pull_requests:*"


class _GitHub:
    """GitHub service permissions"""

    Repos = _GitHubRepos()
    Issues = _GitHubIssues()
    PullRequests = _GitHubPullRequests()
    All: Permission = "github:*:*"


class _SalesforceAccounts:
    """Salesforce Accounts permissions"""

    Read: Permission = "salesforce:accounts:read"
    Write: Permission = "salesforce:accounts:write"
    All: Permission = "salesforce:accounts:*"


class _SalesforceLeads:
    """Salesforce Leads permissions"""

    Read: Permission = "salesforce:leads:read"
    Write: Permission = "salesforce:leads:write"
    All: Permission = "salesforce:leads:*"


class _Salesforce:
    """Salesforce service permissions"""

    Accounts = _SalesforceAccounts()
    Leads = _SalesforceLeads()
    All: Permission = "salesforce:*:*"


class _StripePayments:
    """Stripe Payments permissions"""

    Read: Permission = "stripe:payments:read"


class _StripeCustomers:
    """Stripe Customers permissions"""

    Read: Permission = "stripe:customers:read"


class _StripeInvoices:
    """Stripe Invoices permissions"""

    Read: Permission = "stripe:invoices:read"


class _Stripe:
    """Stripe service permissions (read-only)"""

    Payments = _StripePayments()
    Customers = _StripeCustomers()
    Invoices = _StripeInvoices()
    All: Permission = "stripe:*:*"


class _Permissions:
    """
    Type-safe permission constants for AgentAuth SDK

    Example:
        >>> from agentauth_sdk import Permissions
        >>> permissions = [
        ...     Permissions.Zendesk.Tickets.Read,
        ...     Permissions.Slack.Messages.Write,
        ...     Permissions.Admin,
        ... ]
    """

    Zendesk = _Zendesk()
    Slack = _Slack()
    HubSpot = _HubSpot()
    GitHub = _GitHub()
    Salesforce = _Salesforce()
    Stripe = _Stripe()
    Admin: Permission = "*:*:*"


# Singleton instance
Permissions = _Permissions()

File: agentauth-sdk/src/permissions.ts
/**
 * Type-safe permission system for AgentAuth
 * Provides auto-completion for all service:resource:action permissions
 */

// Service definitions
export type Service = 'zendesk' | 'slack' | 'hubspot' | 'github' | 'salesforce' | 'stripe';

// Resource definitions per service
export type ZendeskResource = 'tickets' | 'users';
export type SlackResource = 'messages' | 'channels';
export type HubSpotResource = 'contacts' | 'deals' | 'companies';
export type GitHubResource = 'repos' | 'issues' | 'pull_requests';
export type SalesforceResource = 'accounts' | 'leads';
export type StripeResource = 'payments' | 'customers' | 'invoices';

// Action definitions
export type Action = 'read' | 'write';

// Combine into specific permission types
export type ZendeskPermission = `zendesk:${ZendeskResource}:${Action}`;
export type SlackPermission = `slack:${SlackResource}:${Action}`;
export type HubSpotPermission = `hubspot:${HubSpotResource}:${Action}`;
export type GitHubPermission = `github:${GitHubResource}:${Action}`;
export type SalesforcePermission = `salesforce:${SalesforceResource}:${Action}`;
export type StripePermission = `stripe:${StripeResource}:read`; // Stripe is read-only

// Wildcard permissions
export type AdminPermission = '*:*:*';
export type ServiceWildcard = `${Service}:*:*`;
export type ResourceWildcard =
  | `zendesk:${ZendeskResource}:*`
  | `slack:${SlackResource}:*`
  | `hubspot:${HubSpotResource}:*`
  | `github:${GitHubResource}:*`
  | `salesforce:${SalesforceResource}:*`
  | `stripe:${StripeResource}:*`;

// Combined permission type with full auto-completion
export type Permission =
  | ZendeskPermission
  | SlackPermission
  | HubSpotPermission
  | GitHubPermission
  | SalesforcePermission
  | StripePermission
  | AdminPermission
  | ServiceWildcard
  | ResourceWildcard;

// Permission builder helpers for better DX
export const Permissions = {
  Zendesk: {
    Tickets: {
      Read: 'zendesk:tickets:read' as const,
      Write: 'zendesk:tickets:write' as const,
      All: 'zendesk:tickets:*' as const,
    },
    Users: {
      Read: 'zendesk:users:read' as const,
      Write: 'zendesk:users:write' as const,
      All: 'zendesk:users:*' as const,
    },
    All: 'zendesk:*:*' as const,
  },
  Slack: {
    Messages: {
      Read: 'slack:messages:read' as const,
      Write: 'slack:messages:write' as const,
      All: 'slack:messages:*' as const,
    },
    Channels: {
      Read: 'slack:channels:read' as const,
      Write: 'slack:channels:write' as const,
      All: 'slack:channels:*' as const,
    },
    All: 'slack:*:*' as const,
  },
  HubSpot: {
    Contacts: {
      Read: 'hubspot:contacts:read' as const,
      Write: 'hubspot:contacts:write' as const,
      All: 'hubspot:contacts:*' as const,
    },
    Deals: {
      Read: 'hubspot:deals:read' as const,
      Write: 'hubspot:deals:write' as const,
      All: 'hubspot:deals:*' as const,
    },
    Companies: {
      Read: 'hubspot:companies:read' as const,
      All: 'hubspot:companies:*' as const,
    },
    All: 'hubspot:*:*' as const,
  },
  GitHub: {
    Repos: {
      Read: 'github:repos:read' as const,
      Write: 'github:repos:write' as const,
      All: 'github:repos:*' as const,
    },
    Issues: {
      Read: 'github:issues:read' as const,
      Write: 'github:issues:write' as const,
      All: 'github:issues:*' as const,
    },
    PullRequests: {
      Read: 'github:pull_requests:read' as const,
      All: 'github:pull_requests:*' as const,
    },
    All: 'github:*:*' as const,
  },
  Salesforce: {
    Accounts: {
      Read: 'salesforce:accounts:read' as const,
      Write: 'salesforce:accounts:write' as const,
      All: 'salesforce:accounts:*' as const,
    },
    Leads: {
      Read: 'salesforce:leads:read' as const,
      Write: 'salesforce:leads:write' as const,
      All: 'salesforce:leads:*' as const,
    },
    All: 'salesforce:*:*' as const,
  },
  Stripe: {
    Payments: {
      Read: 'stripe:payments:read' as const,
    },
    Customers: {
      Read: 'stripe:customers:read' as const,
    },
    Invoices: {
      Read: 'stripe:invoices:read' as const,
    },
    All: 'stripe:*:*' as const,
  },
  Admin: '*:*:*' as const,
} as const;

// Helper to validate permission format
export function isValidPermission(permission: string): permission is Permission {
  const parts = permission.split(':');
  if (parts.length !== 3) return false;

  const [service, resource, action] = parts;

  // Allow wildcards
  if (service === '*' && resource === '*' && action === '*') return true;

  // Check if service is valid
  const validServices = ['zendesk', 'slack', 'hubspot', 'github', 'salesforce', 'stripe'];
  if (!validServices.includes(service) && service !== '*') return false;

  return true;
}
